### PR TITLE
rpc-v2: Enable the `archive` class of methods

### PR DESCRIPTION
**Pull request by lexnv**
The [archive](https://github.com/paritytech/json-rpc-interface-spec/blob/main/src/api/archive.md) API is unstable and subject to change.

This PR enables the `archive` class of the RPC-V2 spec to substrate based chains.

The `archive` API is enabled for archive nodes: 
- the state of the blocks is in archive mode
- the block's bodies are in archive mode

While at it, this PR extends the `BlocksPrunning` enum with an `is_archive` helper to check if the pruning mode keeps the block's bodies for long enough.

Defaults used for the `archive` API:
- a maximum of 5 responses are provided for descendants queries (this is similar to chainHead)
- a maximum of 8 item queries are accepted at a time

Before stabilizing the API we should look into these defaults and adjust after collecting some data.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4994497

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4995475

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4995905

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4995939

File: substrate/client/service/src/builder.rs [hunk 1]
```diff
@@ -664,6 +666,22 @@ where
 	)
 	.into_rpc();
 
+	// Part of the RPC v2 spec.
+	// An archive node that can respond to the `archive` RPC-v2 queries is a node with:
+	// - state pruning in archive mode: The storage of blocks is kept around
+	// - block pruning in archive mode: The block's body is kept around
+	let is_archive_node = config.state_pruning.as_ref().map(|sp| sp.is_archive()).unwrap_or(false) &&
+		config.blocks_pruning.is_archive();
+	let genesis_hash = client.hash(Zero::zero()).ok().flatten().expect("Genesis block exists; qed");
```
Comment (jsdw) [inline substrate/client/service/src/builder.rs:675]: Are things too generic to use `0` here instead of `Zero::zero()`?

**Review by jsdw — APPROVED** [review]

File: substrate/client/service/src/builder.rs (hunk 1 above)
Comment (lexnv) [reply substrate/client/service/src/builder.rs:675]: Yep, the `.hash` needs a ` NumberFor<Block>`; this is similar [here](https://github.com/paritytech/polkadot-sdk/blob/lexnv/enable-archive/substrate/client/service/src/builder.rs#L931)

Comment (niklasad1) [reply substrate/client/service/src/builder.rs:675]: nit: just create this if is_archive_node == true?

File: substrate/client/service/src/builder.rs [hunk 2]
```diff
@@ -664,6 +666,22 @@ where
...
+	let archive_v2 = sc_rpc_spec_v2::archive::Archive::new(
```
[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/client/service/src/builder.rs]: Yep, it makse sense! I've kept this separate to group the `rpc_api.merge` at the end of the function; however I'll merge the archive_v2 here instead 👍

**Review by niklasad1 — APPROVED** [review]

**Pull request merged**